src/install_pypi/mod.rs

```diff
@@ -565,6 +568,99 @@ impl PyPIPrefixUpdater {
         Ok(remote_dists)
     }
 
+    /// Verify hashes for all direct URL distributions
+    async fn verify_all_direct_url_hashes(
+        &self,
+        _local: &[(CachedDist, InstallReason)],
+        remote: &[(Dist, InstallReason)],
+    ) -> miette::Result<()> {
+        future::try_join_all(remote.iter().map(|(dist, _)| self.verify_dist_hash(dist))).await?;
+        Ok(())
+    }
+
+    /// Extract hash info from URL fragment like "#sha256=abc123" or "#md5=def456&egg=foo"
+    fn extract_hash_from_fragment(fragment: &str) -> Option<(&str, &str)> {
+        fragment.split('&').find_map(|param| {
+            let (key, value) = param.split_once('=')?;
+            match key.to_lowercase().as_str() {
+                "sha256" => Some((value, "sha256")),
+                "md5" => Some((value, "md5")),
+                _ => None,
+            }
+        })
+    }
+
+    /// Verify hash for a single distribution
+    async fn verify_dist_hash(&self, dist: &Dist) -> miette::Result<()> {
+        use rattler_digest::{Md5, Sha256, compute_bytes_digest};
+
+        let (url, name) = match dist {
+            Dist::Built(uv_distribution_types::BuiltDist::DirectUrl(direct)) => {
+                (direct.url.to_url(), dist.name())
+            }
+            _ => return Ok(()),
+        };
+
+        let Some((expected_hash, algorithm)) =
+            url.fragment().and_then(Self::extract_hash_from_fragment)
+        else {
+            return Ok(());
+        };
+
+        tracing::info!(
+            "Verifying {} hash for {}: expected {}",
+            algorithm,
+            name,
+            expected_hash
+        );
+
+        // Download file for verification
+        let clean_url = url
+            .as_str()
+            .split('#')
+            .next()
+            .ok_or_else(|| miette::miette!("Invalid URL format"))?;
+        let clean_url = url::Url::parse(clean_url).into_diagnostic()?;
+
+        let bytes = self
+            .registry_client
+            .uncached_client(&clean_url)
+            .get(clean_url.as_str())
+            .send()
+            .await
+            .into_diagnostic()
+            .context("Failed to download distribution for hash verification")?
+            .bytes()
```

Comment:
This will load the entire dist into memory. Please use streaming to compute the hash.

Reply:
Wont this also download every package twice?? Wont this also download the package even if it is already in cache?